core/src/main/scala/dagr/core/tasksystem/Pipe.scala

```diff
@@ -0,0 +1,182 @@
+}
+
+/** A simplified trait for sink tasks that can receive data via a pipe, but cannot pipe onwards. */
+trait PipeIn[In] extends Pipe[In,Nothing]
+
+/** A simplified trait for generate tasks that can pipe out, but cannot recieve data from a pipe. */
+trait PipeOut[Out] extends Pipe[Nothing,Out]
+
+
+///////////////////////////////////////////////////////////////////////////////
+// Section: Objects and private utility classes to make it all work
+///////////////////////////////////////////////////////////////////////////////
+object Pipes {
+  /**
+    * A class that represents an empty pipe, that allows for easier construction of pipes with conditional branches.
+    * Must remain private and only be intantiated by Piping.empty().
+    */
+  private class EmptyPipe[T] extends Pipe[T,T] {
+    override def args: Seq[Any] = throw new IllegalStateException("Should not reach here.")
+    override def pickResources(availableResources: ResourceSet): Option[ResourceSet] = throw new IllegalStateException("Should not reach here.")
+  }
+
+  /**
+    * An immutable class representing a chain of tasks that can be piped together. Currently only supports chains
+    * with up to a maximum of one non-FixedResources task, in order to keep the resource management code simple.
+    *
+    * The constructor does not check types as thoroughly as one might like, instead this is done in
+    * [[Pipes.chain()]], which is the only place the constructor is called from.
+    */
+  private class PipeChain[In,Out](val tasks: List[Pipe[_,_]], val first: Pipe[In,_], val last: Pipe[_,Out]) extends ProcessTask with Pipe[In,Out] {
+    if (tasks.filterNot(_.isInstanceOf[FixedResources]).size > 1) {
+      throw new IllegalArgumentException("Pipe chain includes more than one task that does not implement FixedResources.")
+    }
+
+    override private[tasksystem] def left: Pipe[In, _] = this.first
+    override private[tasksystem] def right: Pipe[_, Out] = this.last
+    override private[tasksystem] def ordered: Seq[Pipe[_, _]] = this.tasks
+
+    /** Returns a single Seq over the args of all piped tasks, with pipe operators between them! */
+    override def args: Seq[Any] = {
+      tasks.foldLeft[List[Any]](Nil)((list, task) => list ++ list.headOption.map(x => "|") ++ task.args)
+    }
+
+    /**
+      * Splits the tasks into a seq of tasks that implement fixed resources, and optionally one
+      * task that does not implement FixedResources.
+      */
+    private def partition: (Seq[FixedResources], Option[ProcessTask]) = {
+      val (fixed, variable) = this.tasks.partition(task => task.isInstanceOf[FixedResources])
+      (fixed.map(_.asInstanceOf[FixedResources]), variable.headOption)
+    }
+
+
+    /**
+      * Picks resources for the piped set of tasks by attempting to allocate the necessary resources
+      * to all the tasks that implement FixedResources, and then offering the rest to the optional
```

Comment:
in comments, if you reference code or classes, could you use "`code`" or `[[ClassName]]`?